src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -218,32 +218,30 @@ Thread::SuspendThreadResult Thread::SuspendThread(BOOL fOneTryOnly, DWORD *pdwSu
 
     while (TRUE) {
         StateHolder<MyEnterLogLock, MyLeaveLogLock> LogLockHolder(FALSE);
-
         CounterHolder handleHolder(&m_dwThreadHandleBeingUsed);
 
-        // Whether or not "goto retry" should YieldProcessor and __SwitchToThread
-        BOOL doSwitchToThread = TRUE;
-
         hThread = GetThreadHandle();
         if (hThread == INVALID_HANDLE_VALUE) {
             str = STR_UnstartedOrDead;
             break;
         }
-        else if (hThread == SWITCHOUT_HANDLE_VALUE) {
-            str = STR_SwitchedOut;
-            break;
-        }
 
         {
             // We do not want to suspend the target thread while it is holding the log lock.
             // By acquiring the lock ourselves, we know that this is not the case.
+            // Note: LogLock is a noop when not logging, do not assume the rest is running under a lock.
             LogLockHolder.Acquire();
 
             // It is important to avoid two threads suspending each other.
             // Before a thread suspends another, it increments its own m_dwForbidSuspendThread count first,
             // then it checks the target thread's m_dwForbidSuspendThread.
             ForbidSuspendThreadHolder forbidSuspend;
-            if ((m_dwForbidSuspendThread != 0))
+
+            // We need a total order between write/read of our/their m_dwForbidSuspendThread here, thus a full fence.
+            // Note - this is just to prevent mutual suspension of two threads executing this same sequence.
+            // The other thread may still set its m_dwForbidSuspendThread for other reasons asynchronously.
+            // We will check on that once the thread is suspended.
+            if (InterlockedOr(&m_dwForbidSuspendThread, 0))
```

Comment:
Another way of looking at this is to consider the following scenario:
- thread A has incremented its forbid count (but the write is still in the cpu write buffer)
- thread B has incremented its forbid count (but the write is still in the cpu write buffer)
- threads A checks the forbid count of thread B.  It does not matter volatile read or not, it can't see the changes which are not yet committed to memory on another core.
- threads B checks the forbid count of thread A.  It does not matter volatile read or not, it can't see the changes which are not yet committed to memory on another core.
- and in some rare case both A and B could proceed to suspend each other at the same time and result in a deadlock. (even on x64)

The part that `m_dwForbidSuspendThread` is `Volatile` is not helping. It is not necessary in other places where we read it from the current thread of from a suspended thread, and it is not sufficient here. When two live threads try to coordinate their forbids, they need to make sure that the reads do not happen earlier than writes.
`InterlockedOr` is a one way to provide that. Full memory fence immediately before the read could be another option.

I did not remove `Volatile` from `m_dwForbidSuspendThread`, though it does not seems very useful.